### DISCUSSION TITLE
Discuss about Memtest86plus runtime occupied memory region

### DISCUSSION THREAD

**Discussion by yoshinoyatoko** [memtest86plus/memtest86plus#260] [General]
Hi
Memtest86plus is a good mem test tool.
It acts as a standalone OS, so it could test more memory range than just as a tronditional uefi shell app.

So, i am curious about its running mechnism, especially for its own occupied memory range during runtime.
Could you explain its running mechanism?

For example:
Does this app usually need how many memory ranges for its normal running?
Could This app's image be rebased during runtime so that to test its prevous occupied memory ranges?

**Comment by debrouxl** [comment]
memtest86+ does indeed relocate itself at runtime, flipping between a location in the 0-4 MB range (currently; up from 0-1 MB after reports about the physical memory map on some motherboards making it impossible to find a consecutive memory block of ~400 KB in the 0-1 MB range) and a location above 4 MB. To achieve this, the binary embeds relocation information, and code to parse and apply these relocations after copying the program's code and data to the new location. The more data structures containing global string or function pointers, the more relocations.
Combined with the call to the UEFI boot services exit function, in order to free the dozens or hundreds of MBs of RAM (per other issues / discussions in this repo) reserved by the UEFI boot services, this makes it possible to test just about the entire non-reserved physical memory range.

**Comment by yoshinoyatoko** [comment]
Hi, debrouxl
Thanks a lot!
because memtest86+ could rebase itself at runtime, so it own previous occupied memory range could aslo be tested after finished rebased?
We dont worry about that memtest86+ runtime owned memory ranges could not be tested.

**Comment by debrouxl** [comment]
> because memtest86+ could rebase itself at runtime, so it own previous occupied memory range could aslo be tested after finished rebased?

Yes, exactly.

**Comment by yoshinoyatoko** [comment]
Thanks a lot.
